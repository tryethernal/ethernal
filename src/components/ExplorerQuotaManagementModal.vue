<template>
    <v-dialog v-model="dialog" max-width="1200" :persistent="true">
        <v-card outlined>
            <v-card-title>
                <template>Transaction Quota Management</template>
                <v-spacer></v-spacer>
                <v-btn icon @click="close()" ><v-icon>mdi-close</v-icon></v-btn>
            </v-card-title>
            <v-card-text>
                <v-alert text type="error" v-if="errorMessage">{{ errorMessage }}</v-alert>
                <v-alert text type="success" v-if="successMessage">{{ successMessage }}</v-alert>
                <p>
                    Your plan includes {{ baseQuota.toLocaleString() }} transactions per month. If you need more, you can extend your quota here.
                    If set, your quota extension will automatically be renewed every month, but you can adjust or remove it anytime.
                    Increases will be charged for right away, and decreases will result in a prorated credit applied to your future invoices.
                </p>
                <b>Current Quota:</b> {{ baseQuota.toLocaleString() }} base + {{ currentExtraQuota.toLocaleString() }} extra = {{ currentQuota.toLocaleString() }} txs
                <v-divider class="mt-2 mb-4"></v-divider>
                <v-skeleton-loader v-if="stripePlanLoading" type="list-item-three-line"></v-skeleton-loader>
                <v-form v-else @submit.prevent="updateQuotaExtension()" v-model="valid">
                    <v-text-field
                        :rules="[v => parseInt(v) == 0 || parseInt(v) >= 10000 || 'Cannot be less than 10,000.']"
                        dense
                        outlined
                        v-model="rawExtraQuota"
                        type="number"
                        suffix="extra transactions per month"
                        label="Extra Transaction Quota"></v-text-field>
                    <v-simple-table>
                        <template v-slot:default>
                            <thead>
                                <tr>
                                    <th class="text-left">Up to</th>
                                    <th class="text-left">Cost per tx</th>
                                    <th class="text-left">Number of txs</th>
                                    <th class="text-left">Total Monthly Cost</th>
                                </tr>
                            </thead>
                            <tbody>
                                <tr v-for="(range, idx) in ranges" :key="idx">
                                    <td>{{ range.upTo.toLocaleString() }} {{ typeof range.upTo == 'number' ? 'txs' : '' }}</td>
                                    <td>${{ range.cost }}</td>
                                    <td>{{ quantityForRange(idx).toLocaleString() }}</td>
                                    <td>${{ costForRange(idx).toLocaleString() }}</td>
                                </tr>
                            </tbody>
                            <tfoot>
                                <tr>
                                    <td></td>
                                    <td></td>
                                    <td class="font-weight-bold">{{ totalQuantity.toLocaleString() }}</td>
                                    <td class="font-weight-bold">${{ totalCost.toLocaleString() }}</td>
                                </tr>
                            </tfoot>
                        </template>
                    </v-simple-table>
                    <div align="right">
                        <v-btn color="primary" type="submit" :loading="loading || stripePlanLoading" :disabled="!valid">Update Quota</v-btn>
                    </div>
                </v-form>
            </v-card-text>
        </v-card>
    </v-dialog>
</template>
<script>
const moment = require('moment');

export default {
    name: 'ExplorerQuotaManagementModal',
    data: () => ({
        dialog: false,
        resolve: null,
        reject: null,
        rawExtraQuota: null,
        subscription: {},
        explorerId: null,
        valid: false,
        loading: false,
        stripePlanLoading: false,
        stripePlan: null,
        errorMessage: null,
        successMessage: null
    }),
    methods: {
        moment,
        open(options) {
            this.dialog = true;
            this.stripePlanLoading = true;
            this.subscription = options.subscription;
            this.explorerId = options.explorerId;
            if (this.subscription.stripeQuotaExtension)
                this.rawExtraQuota = this.subscription.stripeQuotaExtension.quota;

            this.server.getQuotaExtensionPlan()
                .then(({ data }) => this.stripePlan = data)
                .finally(() => this.stripePlanLoading = false);

            return new Promise((resolve, reject) => {
                this.resolve = resolve;
                this.reject = reject;
            });
        },
        close() {
            this.resolve();
            this.dialog = false;
        },
        quantityForRange(idx) {
            const lowerBound = idx > 0 ? this.ranges[idx - 1].upTo : 0;
            const upperBound = this.ranges[idx].upTo;
            const quotaToUse = this.valid ? this.newQuota : this.currentQuota;

            if (idx == 0)
                return Math.min(quotaToUse, this.ranges[idx].upTo)
            else if (idx > 0 && idx < this.ranges.length - 1)
                return Math.max(Math.min(quotaToUse - lowerBound, upperBound - lowerBound), 0);
            else
                return Math.max(quotaToUse - this.ranges[idx - 1].upTo, 0);
        },
        costForRange(idx) {
            return parseFloat(this.quantityForRange(idx)) * parseFloat(this.ranges[idx].cost);
        },
        updateQuotaExtension() {
            this.loading = true;
            this.successMessage = null;
            this.errorMessage = null;

            if (parseInt(this.extraQuota) == 0)
                this.server.cancelQuotaExtension(this.explorerId)
                    .then(({ data: { stripeSubscription }}) => {
                        this.successMessage = 'Transaction quota updated.';
                        this.subscription = stripeSubscription;
                    })
                    .catch(error => {
                        this.errorMessage = error.response && error.response.data;
                    })
                    .finally(() => {
                        this.loading = false;
                    });
            else
                this.server.updateQuotaExtension(this.explorerId, this.stripePlan.slug, this.extraQuota)
                    .then(({ data: { stripeSubscription }}) => {
                        this.successMessage = 'Transaction quota updated.';
                        this.subscription = stripeSubscription;
                    })
                    .catch(error => {
                        this.errorMessage = error.response && error.response.data;
                    })
                    .finally(() => {
                        this.loading = false;
                    });
        }
     },
    computed: {
        ranges() {
            return this.stripePlan ? this.stripePlan.capabilities.ranges : [];
        },
        totalQuantity() {
            return this.ranges.reduce((acc, _curr, idx) => acc + this.quantityForRange(idx), 0);
        },
        totalCost() {
            return this.ranges.reduce((acc, _curr, idx) => acc + this.costForRange(idx), 0);
        },
        baseQuota() {
            return parseInt(this.subscription.stripePlan && this.subscription.stripePlan.capabilities.txLimit || 0);
        },
        extraQuota() {
            return this.rawExtraQuota && parseInt(this.rawExtraQuota) || 0;
        },
        newQuota() {
            return this.baseQuota + this.extraQuota;
        },
        currentExtraQuota() {
            return this.subscription.stripeQuotaExtension ? this.subscription.stripeQuotaExtension.quota : 0;
        },
        currentQuota() {
            return this.baseQuota + this.currentExtraQuota;
        }
    }
}
</script>
