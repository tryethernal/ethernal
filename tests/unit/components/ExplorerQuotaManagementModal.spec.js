import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import ExplorerQuotaManagementModal from '@/components/ExplorerQuotaManagementModal.vue';

let helper;
beforeEach(() => {
    jest.clearAllMocks()
    helper = new MockHelper();
});

const stripePlanData = {
    slug: 'transaction-quota',
    capabilities: {"quotaExtension": true, "ranges": [{"upTo": 100000, "cost": 0.005}, {"upTo": 1000000, "cost":0.0007}, { "upTo": 10000000, "cost": 0.00015}, {"upTo": "âˆž", "cost": 0.000045}]},
}

describe('ExplorerQuotaManagementModal.vue', () => {
    it('Should display current quota info', async () => {
        const wrapper = helper.mountFn(ExplorerQuotaManagementModal);
        jest.spyOn(helper.mocks.server, 'getQuotaExtensionPlan').mockResolvedValueOnce({ data: stripePlanData });

        wrapper.vm.open({
            subscription: {
                explorerId: 1,
                stripePlan: {
                    capabilities: { txLimit: 100000 }
                },
                stripeQuotaExtension: {
                    quota: 20000
                }
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display base cost simulation', async () => {
        const wrapper = helper.mountFn(ExplorerQuotaManagementModal);
        jest.spyOn(helper.mocks.server, 'getQuotaExtensionPlan').mockResolvedValueOnce({ data: stripePlanData });

        wrapper.vm.open({
            subscription: {
                explorerId: 1,
                stripePlan: {
                    capabilities: { txLimit: 100000 }
                },
                stripeQuotaExtension: {
                    quota: 20000
                }
            }
        });
        await flushPromises();

        wrapper.setData({ rawExtraQuota: 4000000 });
        await wrapper.vm.$nextTick();

        expect(wrapper.html()).toMatchSnapshot();
    });
});
