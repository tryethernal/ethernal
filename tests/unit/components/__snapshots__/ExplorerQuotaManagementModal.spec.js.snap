// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExplorerQuotaManagementModal.vue Should display base cost simulation 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active v-dialog--persistent" style="transform-origin: center center; max-width: 1200px;">
        <div class="v-card v-sheet v-sheet--outlined theme--light">
          <div class="v-card__title">Transaction Quota Management <div class="spacer"></div> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-close theme--light"></i></span></button></div>
          <div class="v-card__text">
            <!---->
            <!---->
            <p>
              Your plan includes 100,000 transactions per month. If you need more, you can extend your quota here.
              If set, your quota extension will automatically be renewed every month, but you can adjust or remove it anytime.
              Increases will be charged for right away, and decreases will result in a prorated credit applied to your future invoices.
            </p> <b>Current Quota:</b> 100,000 base + 20,000 extra = 120,000 txs
            <hr role="separator" aria-orientation="horizontal" class="mt-2 mb-4 v-divider theme--light">
            <form novalidate="novalidate" class="v-form">
              <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="input-33" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Extra Transaction Quota</label><input id="input-33" type="number">
                      <div class="v-text-field__suffix">extra transactions per month</div>
                    </div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
              <div class="v-data-table theme--light">
                <div class="v-data-table__wrapper">
                  <table>
                    <thead>
                      <tr>
                        <th class="text-left">Up to</th>
                        <th class="text-left">Cost per tx</th>
                        <th class="text-left">Number of txs</th>
                        <th class="text-left">Total Monthly Cost</th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr>
                        <td>100,000 txs</td>
                        <td>$0.005</td>
                        <td>100,000</td>
                        <td>$500</td>
                      </tr>
                      <tr>
                        <td>1,000,000 txs</td>
                        <td>$0.0007</td>
                        <td>900,000</td>
                        <td>$630</td>
                      </tr>
                      <tr>
                        <td>10,000,000 txs</td>
                        <td>$0.00015</td>
                        <td>3,100,000</td>
                        <td>$465</td>
                      </tr>
                      <tr>
                        <td>∞ </td>
                        <td>$0.000045</td>
                        <td>0</td>
                        <td>$0</td>
                      </tr>
                    </tbody>
                    <tfoot>
                      <tr>
                        <td></td>
                        <td></td>
                        <td class="font-weight-bold">4,100,000</td>
                        <td class="font-weight-bold">$1,595</td>
                      </tr>
                    </tfoot>
                  </table>
                </div>
              </div>
              <div align="right"><button type="submit" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Update Quota</span></button></div>
            </form>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ExplorerQuotaManagementModal.vue Should display current quota info 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active v-dialog--persistent" style="transform-origin: center center; max-width: 1200px;">
        <div class="v-card v-sheet v-sheet--outlined theme--light">
          <div class="v-card__title">Transaction Quota Management <div class="spacer"></div> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-close theme--light"></i></span></button></div>
          <div class="v-card__text">
            <!---->
            <!---->
            <p>
              Your plan includes 100,000 transactions per month. If you need more, you can extend your quota here.
              If set, your quota extension will automatically be renewed every month, but you can adjust or remove it anytime.
              Increases will be charged for right away, and decreases will result in a prorated credit applied to your future invoices.
            </p> <b>Current Quota:</b> 100,000 base + 20,000 extra = 120,000 txs
            <hr role="separator" aria-orientation="horizontal" class="mt-2 mb-4 v-divider theme--light">
            <form novalidate="novalidate" class="v-form">
              <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <fieldset aria-hidden="true">
                      <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="input-14" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Extra Transaction Quota</label><input id="input-14" type="number">
                      <div class="v-text-field__suffix">extra transactions per month</div>
                    </div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
              <div class="v-data-table theme--light">
                <div class="v-data-table__wrapper">
                  <table>
                    <thead>
                      <tr>
                        <th class="text-left">Up to</th>
                        <th class="text-left">Cost per tx</th>
                        <th class="text-left">Number of txs</th>
                        <th class="text-left">Total Monthly Cost</th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr>
                        <td>100,000 txs</td>
                        <td>$0.005</td>
                        <td>100,000</td>
                        <td>$500</td>
                      </tr>
                      <tr>
                        <td>1,000,000 txs</td>
                        <td>$0.0007</td>
                        <td>20,000</td>
                        <td>$14</td>
                      </tr>
                      <tr>
                        <td>10,000,000 txs</td>
                        <td>$0.00015</td>
                        <td>0</td>
                        <td>$0</td>
                      </tr>
                      <tr>
                        <td>∞ </td>
                        <td>$0.000045</td>
                        <td>0</td>
                        <td>$0</td>
                      </tr>
                    </tbody>
                    <tfoot>
                      <tr>
                        <td></td>
                        <td></td>
                        <td class="font-weight-bold">120,000</td>
                        <td class="font-weight-bold">$514</td>
                      </tr>
                    </tfoot>
                  </table>
                </div>
              </div>
              <div align="right"><button type="submit" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Update Quota</span></button></div>
            </form>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
